go/vt/vtgate/executor.go

```diff
@@ -304,6 +304,113 @@ func (e *Executor) handleRollback(ctx context.Context, safeSession *SafeSession,
 	return &sqltypes.Result{}, err
 }
 
+func (e *Executor) handleSavepoint(ctx context.Context, safeSession *SafeSession, sql string, logStats *LogStats) (*sqltypes.Result, error) {
+	execStart := time.Now()
+	logStats.PlanTime = execStart.Sub(logStats.StartTime)
+	logStats.ShardQueries = uint32(len(safeSession.ShardSessions))
+	e.updateQueryCounts("Savepoint", "", "", int64(logStats.ShardQueries))
+	defer func() {
+		logStats.ExecuteTime = time.Since(execStart)
+	}()
+
+	if len(safeSession.ShardSessions) == 0 {
+		if safeSession.InTransaction() {
+			// Storing savepoint as this needs to be executed just after starting transaction.
+			safeSession.StoreSavepoint(sql)
+			return &sqltypes.Result{}, nil
+		}
+		// Safely to ignore as there is no transaction.
+		return &sqltypes.Result{}, nil
+	}
+	var rss []*srvtopo.ResolvedShard
+	for _, shardSession := range safeSession.ShardSessions {
+		rss = append(rss, &srvtopo.ResolvedShard{
+			Target:  shardSession.Target,
+			Gateway: e.resolver.resolver.GetGateway(),
+		})
+	}
+	queries := make([]*querypb.BoundQuery, len(rss))
+	for i := range rss {
+		queries[i] = &querypb.BoundQuery{Sql: sql}
+	}
+
+	qr, errs := e.ExecuteMultiShard(ctx, rss, queries, safeSession, false, false)
+	err := vterrors.Aggregate(errs)
+	if err != nil {
+		return nil, err
+	}
+	safeSession.StoreSavepoint(sql)
+	return qr, nil
+}
+
+func (e *Executor) handleSRollback(ctx context.Context, safeSession *SafeSession, sql string, logStats *LogStats) (*sqltypes.Result, error) {
+	execStart := time.Now()
+	logStats.PlanTime = execStart.Sub(logStats.StartTime)
+	logStats.ShardQueries = uint32(len(safeSession.ShardSessions))
+	e.updateQueryCounts("Rollback Savepoint", "", "", int64(logStats.ShardQueries))
+	defer func() {
+		logStats.ExecuteTime = time.Since(execStart)
+	}()
+
+	if len(safeSession.ShardSessions) == 0 {
+		return nil, mysql.NewSQLError(mysql.ERSavepointNotExist, "42000", "SAVEPOINT does not exist: %s", sql)
+	}
```

Comment:
I'd say we should play it safe and add this also to the list of savepoint statements.

This also means that three functions can be collapsed into a single parameterized one?

Reply:
I tried to imitated the mysql behaviour. https://dev.mysql.com/doc/refman/8.0/en/savepoint.html
But, I know where it fails. 
I will add a test for that case: 
`begin;
savepoint a;
rollback to a;`

I can add the same check of if safeSession.InTransaction() { }  all the functions looks same then. I will merge them. good point.

Reply:
Also, we should not store any savepoint contructs till a start transaction is requested or autocommit is false which for us means session.InTransaction to be true.